app/components/current-page.ts

```diff
@@ -0,0 +1,99 @@
+import type Store from '@ember-data/store';
+import type Book from 'collection-demo/models/book';
+import { tracked } from '@glimmer/tracking';
+import type { Collection } from '@ember-data/types';
+import { resource, resourceFactory } from 'ember-resources';
+
+export class AsyncContent<T> {
+  @tracked content: T | undefined;
+  @tracked isLoading = true;
+  @tracked error: string | Error | unknown | null = null;
+
+  get isInitiallyLoading() {
+    return this.isLoading && this.content === undefined;
+  }
+
+  get hasError() {
+    return this.error !== null;
+  }
+}
+
+interface FirstPageOptions {
+  url: string;
+  sort?: string | null;
+  filter?: string | null;
+  genre?: string | null;
+  author?: string | null;
+  page?: number | null;
+  limit?: number | null;
+}
+
+// TODO: inline to ember-resources?
+//  https://github.com/NullVoxPopuli/ember-resources/issues/945
+const unwrap = <T>(x: T | (() => T)): T =>
+  typeof x === 'function' ? (x as any)() : x;
+
+export const CurrentPage = resourceFactory(
+  /**
+   * This way of passing args has nothing to do with resources,
+   * but is just how you pass args for lazy evaluation without creating
+   * an options class with its own tracked properties
+   */
+  (optionsFns: {
+    currentUrl: () => string | null;
+    firstPageOptions: () => FirstPageOptions;
+    addPage: (content: Collection<unknown>) => void;
+  }) => {
+    let previousFirstPageOptions: FirstPageOptions | undefined;
+
+    /**
+     * TODO: add request cancellation
+     *     See example: https://ember-resources.pages.dev/funcs/util_remote_data.RemoteData
+     */
+    return resource(({ on, owner, use }) => {
```

Comment:
on is not used here, as there is no cleanup / request cancellation occurring.

use is not used here as there are no other resources being consumed -- but if `Request` was a resource from ember-data, that could be used